### Pandas:-
Pandas is a Python library used for working with data sets.

It has functions for analyzing 
cleaning, exploring, and manipulating data.

The name "Pandas" has a reference to oth 
"Panel Data", and "Python Data Ana.lysis"

### We Can install, import and upgrade Pandas like:
!pip install pandas

import pandas

import pandas as pd

!pip install --upgrade panda"

### Check the Version of Pandas:-

In [3]:
print(pd.__version__)

2.2.2


### Why Use Pandas?

Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

### Data Science:-

Data Science is a branch of computer science where we study how to store, use and analyze data for deriving information from it.

### What Can Pandas Do?

Pandas gives you answers about the data. Like:

Is there a correlation between two or more columns?
What is average value?
Max value?
Min value?
Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

### We have many functions in Pandas. Now, we discuss its example.

In [2]:
import pandas as pd
mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}
myvar = pd.DataFrame(mydataset)
print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


### Load Files Into a DataFrame:-

If your data sets are stored in a file, Pandas can load them into a DataFrame.

### Read CSV Files:-

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In [27]:
# In our examples we will be using a CSV file called 'Pandas/data.csv'.

# Load the CSV into a DataFrame:
df = pd.read_csv('Pandas/data.csv')
print(df.to_string())

    Duration  Pulse  Maxpulse  Calories
0          A    110       130     409.1
1         60    117       145     479.0
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
5         60    102       127     300.0
6         60    110       136     374.0
7         45    104       134     253.3
8         30    109       133     195.1
9         60     98       124     269.0
10        60    103       147     329.3
11        60    100       120     250.7
12        60    106       128     345.3
13        60    104       132     379.3
14        60     98       123     275.0
15        60     98       120     215.2
16        60    100       120     300.0
17        45     90       112       NaN
18        60    103       123     323.0
19        45     97       125     243.0
20        60    108       131     364.2
21        45    100       119     282.0
22        60    130       101     300.0
23        45    105       132     246.0


### NOTE:- 

If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:

In [14]:
data = pd.read_csv('Pandas/data.csv')
print(data)

    Duration  Pulse  Maxpulse  Calories
0          A    110       130     409.1
1         60    117       145     479.0
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
..       ...    ...       ...       ...
164       60    105       140     290.8
165       60    110       145     300.0
166       60    115       145     310.2
167       75    120       150     320.4
168       75    125       150     330.4

[169 rows x 4 columns]


In [15]:
type(data)

pandas.core.frame.DataFrame

### What is a DataFrame?

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

Data sets in Pandas are usually multi-dimensional tables, called DataFrames.

Series is like a column, a DataFrame is the whole table.


In [21]:
dataset = {
    "Cars" : ["BMW", "Tesla", "Grandi"],
    "Model" : [2015, 2021, 2017]
}
myvar = pd.DataFrame(dataset)
print(myvar)

     Cars  Model
0     BMW   2015
1   Tesla   2021
2  Grandi   2017


In [24]:
dataset3 = {
    "Cars" : ["BMW", "Volvo", "Tesla"],
    "Model" : [2004, 2006, 2021]
}
df = pd.DataFrame(dataset3)
print(df)

    Cars  Model
0    BMW   2004
1  Volvo   2006
2  Tesla   2021


In [16]:
# Create a DataFrame from two Series:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}
myvar = pd.DataFrame(data)
print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45


### Locate Row:-

As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s).

In [21]:
# Example
# Return row 0:
#refer to the row index:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [26]:
school = {
    "8TH" : [80, 90, 99, 100],
    "7TH" : [89, 76, 89, 98]
}
df = pd.DataFrame(school)
print(df)

   8TH  7TH
0   80   89
1   90   76
2   99   89
3  100   98


### Viewing the Data:-

One of the most used method for getting a quick overview of the DataFrame, is the head() method.

The head() method returns the headers and a specified number of rows, starting from the top.

### Note: 

If the number of rows is not specified, the head() method will return the top 5 rows.

In [25]:
data = pd.read_csv('Pandas/data.csv')
data.head() # yaha first 5 rows show ho ge

Duration  Pulse  Maxpulse  Calories
0        A    110       130     409.1
1       60    117       145     479.0
2       60    103       135     340.0
3       45    109       175     282.4
4       45    117       148     406.0

In [26]:
data.head(20) # yaha hum data ki top 20 rows or column ly rhy hn.

Duration  Pulse  Maxpulse  Calories
0         A    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0
10       60    103       147     329.3
11       60    100       120     250.7
12       60    106       128     345.3
13       60    104       132     379.3
14       60     98       123     275.0
15       60     98       120     215.2
16       60    100       120     300.0
17       45     90       112       NaN
18       60    103       123     323.0
19       45     97       125     243.0

In [27]:
data.head(3)

Duration  Pulse  Maxpulse  Calories
0        A    110       130     409.1
1       60    117       145     479.0
2       60    103       135     340.0

In [28]:
data.tail() # tail funtion ma data ki last 5 values aati hn.

Duration  Pulse  Maxpulse  Calories
164       60    105       140     290.8
165       60    110       145     300.0
166       60    115       145     310.2
167       75    120       150     320.4
168       75    125       150     330.4

In [29]:
data.tail(10) # here last 10 values shows

Duration  Pulse  Maxpulse  Calories
159       30     80       120     240.9
160       30     85       120     250.4
161       45     90       130     260.4
162       45     95       130     270.0
163       45    100       140     280.9
164       60    105       140     290.8
165       60    110       145     300.0
166       60    115       145     310.2
167       75    120       150     320.4
168       75    125       150     330.4

### Info About the Data:-

The DataFrames object has a method called info(), that gives you more information about the data set.

In [30]:
# info function():
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    object 
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(2), object(1)
memory usage: 5.4+ KB


### Null Values:-
The info() method also tells us how many Non-Null values there are present in each column, and in our data set it seems like 
there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. 

This is a step towards what is called cleaning data, and you will learn more about that in the next chapters.

### Data Cleaning:-

Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells
Data in wrong format
Wrong data
Duplicates

### Empty Cells:-

Empty cells can potentially give you a wrong result when you analyze data.

### Remove Rows:-
    
One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

In [30]:
# Example:
# Return a new Data Frame with no empty cells:
df = pd.read_csv('Pandas/data.csv')
new_df = df.dropna()
print(new_df.to_string())

# Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

    Duration  Pulse  Maxpulse  Calories
0          A    110       130     409.1
1         60    117       145     479.0
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
5         60    102       127     300.0
6         60    110       136     374.0
7         45    104       134     253.3
8         30    109       133     195.1
9         60     98       124     269.0
10        60    103       147     329.3
11        60    100       120     250.7
12        60    106       128     345.3
13        60    104       132     379.3
14        60     98       123     275.0
15        60     98       120     215.2
16        60    100       120     300.0
18        60    103       123     323.0
19        45     97       125     243.0
20        60    108       131     364.2
21        45    100       119     282.0
22        60    130       101     300.0
23        45    105       132     246.0
24        60    102       126     334.5


### If you want to change the original DataFrame, use the inplace = True argument:

In [31]:
# Example
# Remove all rows with NULL values:
df = pd.read_csv('Pandas/data.csv')
df.dropna(inplace = True)
print(df.to_string())

# Note: Now, the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.

    Duration  Pulse  Maxpulse  Calories
0          A    110       130     409.1
1         60    117       145     479.0
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
5         60    102       127     300.0
6         60    110       136     374.0
7         45    104       134     253.3
8         30    109       133     195.1
9         60     98       124     269.0
10        60    103       147     329.3
11        60    100       120     250.7
12        60    106       128     345.3
13        60    104       132     379.3
14        60     98       123     275.0
15        60     98       120     215.2
16        60    100       120     300.0
18        60    103       123     323.0
19        45     97       125     243.0
20        60    108       131     364.2
21        45    100       119     282.0
22        60    130       101     300.0
23        45    105       132     246.0
24        60    102       126     334.5


### Replace Empty Values:-

Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value.

In [4]:
# Replace NULL values with the number 130:

df = pd.read_csv('Pandas/data.csv')
df.fillna(130, inplace = True) # inplace = True use krny sy file ma permanent changing aa jati h.

In [16]:
df = pd.read_csv('Pandas/data.csv')
df.fillna(130, inplace = False) # yaha permanent changing nhe ay ge.

Duration  Pulse  Maxpulse  Calories
0          A    110       130     409.1
1         60    117       145     479.0
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
..       ...    ...       ...       ...
164       60    105       140     290.8
165       60    110       145     300.0
166       60    115       145     310.2
167       75    120       150     320.4
168       75    125       150     330.4

[169 rows x 4 columns]

### Replace Using Mean, Median, or Mode:-

A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column.

In [8]:
# Calculate the MEAN, and replace any empty values with it
df = pd.read_csv('Pandas/data.csv')
x = df["Calories"].mean()
df["Calories"].fillna(x, inplace = False)

0      409.1
1      479.0
2      340.0
3      282.4
4      406.0
       ...  
164    290.8
165    300.0
166    310.2
167    320.4
168    330.4
Name: Calories, Length: 169, dtype: float64

### What is a Series?

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [3]:
# Example:
# Create a simple Pandas Series from a list:

import pandas as pd
a = [1.5, 7, 2, 2.8, 9, 10.1]
myvar = pd.Series(a)
print(myvar)

0     1.5
1     7.0
2     2.0
3     2.8
4     9.0
5    10.1
dtype: float64


### Labels:-

If nothing else is specified, the values are labeled with their index number. 
    
First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [5]:
# Example
# Return the first value of the Series:
a = [1, 7, 2]
myvar = pd.Series(a)
print(myvar[0])

1


In [6]:
b = ["Adeel", "Ahmed", "MRK"]
myvar = pd.Series(b)
print(myvar)

0    Adeel
1    Ahmed
2      MRK
dtype: object


### Create Labels:- 

With the index argument, you can name your own labels.

In [7]:
# Example
# Create your own labels:
a = [1, 7, 2]
myvar = pd.Series(a, index = ["x", "y", "z"])
print(myvar)

x    1
y    7
z    2
dtype: int64


### Named Indexes:-
With the index argument, you can name your own indexes.

In [9]:
b = [10, 12, 14, 16, 18, 20, 22, 24]
myvar = pd.Series(b, index = ["a", "b", "c", "d", "e", "f", "g", "h"])
print(myvar)

a    10
b    12
c    14
d    16
e    18
f    20
g    22
h    24
dtype: int64


In [10]:
c = ["a", "b", "c", "d", "e", "f", "g", "h"]
myvar = pd.Series(c, index = [10, 12, 14, 16, 18, 20, 22, 24])
print(myvar)

10    a
12    b
14    c
16    d
18    e
20    f
22    g
24    h
dtype: object


### Key/Value Objects as Series:

You can also use a key/value object, like a dictionary, when creating a Series.

In [11]:
# Example
# Create a simple Pandas Series from a dictionary:
calories = {"day1": 420, "day2": 380, "day3": 390}
myvar = pd.Series(calories)
print(myvar)
# Note: The keys of the dictionary become the labels.

day1    420
day2    380
day3    390
dtype: int64


In [12]:
days = {"Sunday" : 1, "Monday" : 2, "Tuesday" : 3, "Thursday" : 4, "Friday" : 5}
myvar = pd.Series(days)
print(myvar)

Sunday      1
Monday      2
Tuesday     3
Thursday    4
Friday      5
dtype: int64


In [15]:
days = {"Sunday" : 1, "Monday" : 2, "Tuesday" : 3, "Thursday" : 4, "Friday" : 5}
myvar = pd.Series(days)
myvar

Sunday      1
Monday      2
Tuesday     3
Thursday    4
Friday      5
dtype: int64

### max_rows:-

The number of rows returned is defined in Pandas option settings.

You can check your system's maximum rows with the pd.options.display.max_rows statement.

In [28]:
# Example
# Check the number of maximum returned rows:
print(pd.options.display.max_rows)

60


In [29]:
# Increase the maximum number of rows to display the entire DataFrame:
pd.options.display.max_rows = 9999
df = pd.read_csv('Pandas/data.csv')
print(df) 

    Duration  Pulse  Maxpulse  Calories
0          A    110       130     409.1
1         60    117       145     479.0
2         60    103       135     340.0
3         45    109       175     282.4
4         45    117       148     406.0
5         60    102       127     300.0
6         60    110       136     374.0
7         45    104       134     253.3
8         30    109       133     195.1
9         60     98       124     269.0
10        60    103       147     329.3
11        60    100       120     250.7
12        60    106       128     345.3
13        60    104       132     379.3
14        60     98       123     275.0
15        60     98       120     215.2
16        60    100       120     300.0
17        45     90       112       NaN
18        60    103       123     323.0
19        45     97       125     243.0
20        60    108       131     364.2
21        45    100       119     282.0
22        60    130       101     300.0
23        45    105       132     246.0


In [41]:
myvar = data["duration"]
print(myvar)

[50, 40, 45]


In [45]:
data["duration"]

[50, 40, 45]

### Wrong Data:-

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

### Replacing Values:-

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7.

In [12]:
# Example:
df.loc[171, "Maxpulse"] = 45

In [13]:
df.tail()

Duration  Pulse  Maxpulse  Calories
165       60  110.0     145.0     300.0
166       60  115.0     145.0     310.2
167       75  120.0     150.0     320.4
168       75  125.0     150.0     330.4
171      NaN    NaN      45.0       NaN

### NOTE:-

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any 
values that are outside of the boundaries.

In [14]:
for x in df.index:
    if df.loc[x, "Pulse"] > 120:
        df.loc[x, "Pulse"] = 120

In [15]:
print(df.to_string())

    Duration  Pulse  Maxpulse  Calories
0          A  110.0     130.0     409.1
1         60  117.0     145.0     479.0
2         60  103.0     135.0     340.0
3         45  109.0     175.0     282.4
4         45  117.0     148.0     406.0
5         60  102.0     127.0     300.0
6         60  110.0     136.0     374.0
7         45  104.0     134.0     253.3
8         30  109.0     133.0     195.1
9         60   98.0     124.0     269.0
10        60  103.0     147.0     329.3
11        60  100.0     120.0     250.7
12        60  106.0     128.0     345.3
13        60  104.0     132.0     379.3
14        60   98.0     123.0     275.0
15        60   98.0     120.0     215.2
16        60  100.0     120.0     300.0
17        45   90.0     112.0       NaN
18        60  103.0     123.0     323.0
19        45   97.0     125.0     243.0
20        60  108.0     131.0     364.2
21        45  100.0     119.0     282.0
22        60  120.0     101.0     300.0
23        45  105.0     132.0     246.0
